File: backend/src/controller/game_server_message.rs
use serde::Serialize;

use crate::{
    controller::dto::{GroupDto, GroupedStonesDto},
    domain::{coord::Coord, group::GroupId, stone_placed_changes::StonePlacedChanges},
};

#[derive(Debug, Serialize)]
#[serde(tag = "type")]
pub enum GameServerMessage {
    BoardData {
        grouped_stones: Vec<GroupedStonesDto>,
    },

    StonePlaced {
        coord: Coord,
        assigned_group: GroupDto,
        captured_groups_ids: Vec<GroupId>,
        merged_groups_ids: Vec<GroupId>,
    },

    InvalidMove {
        coord: Coord,
    }
}

impl From<StonePlacedChanges> for GameServerMessage {
    fn from(value: StonePlacedChanges) -> Self {
        Self::StonePlaced {
            coord: value.coord,
            assigned_group: value.assigned_group.into(),
            captured_groups_ids: value.captured_groups_ids,
            merged_groups_ids: value.merged_groups_ids,
        }
    }
}

File: webapp/src/Gateway/GameServerMessage.ts
import { Coord } from "../Domain/Coord";
import { GroupId } from "../Domain/Group";
import { GroupDto } from "./GroupDto";
import { GroupedStonesDto } from "./GroupedStonesDto";

export type GameServerMessage = BoardDataMessage | StonePlacedMessage;

export interface BoardDataMessage {
  type: "BoardData";
  grouped_stones: Array<GroupedStonesDto>;
}

export interface StonePlacedMessage {
  type: "StonePlaced";
  coord: Coord;
  assigned_group: GroupDto;
  captured_groups_ids: Array<GroupId>;
  merged_groups_ids: Array<GroupId>;
}
